### ISSUE TITLE
playlog() and JSON parsing/imporing failing on UTF-8 issue

### ISSUE THREAD

**Issue by gilou** [savonet/liquidsoap#2790]
**Describe the bug**
I'm using playlog() with persistency, and some files have weird chars (probably ISO-8859-1 chars)  in them. This seems to generate a runtime error on a seemingly valid JSON file, because metadata don't seem to be recoded when written.

The bug might be "bigger" than playlog, but I wouldn't know for sure.

**To Reproduce**
have a file named with a latin1 char in the files/ dir. One triggering the issue was called "L'âge idiot.mp3" encoded in latin1.

then…
```
p = playlog(persistency="/tmp/test.log")

def check(m)
  rm = request.metadata(m)
  p.add(rm)
  true
end

pl = playlist("files", check_next=check)
output.dummy(pl, fallible=true)
```
Let the file be played, stored in log, stop the script, restart, let it load, then…
```
Error 14: Uncaught runtime error:
type: json,
message: "Parse error: Sedlexing.MalFormed",
stack: at /usr/share/liquidsoap/2.1.2/libs/utils.liq, line 60, char 6-92,
at /usr/share/liquidsoap/2.1.2/libs/utils.liq, line 59 char 4 - line 64 char 7,
at /usr/share/liquidsoap/2.1.2/libs/utils.liq, line 58 char 2 - line 65 char 5,
at test_pl.liq, line 1, char 4-40

```

**Expected behavior**
Not sure, not giving a runtime error? I'm not sure if I can actually catch that. I have a "solution" that might be a fix idea for playlog: if I recode all metadata in UTF-8, it seems I can have it not crash, and still match using the default hash… The log "looks" ugly then¹…

```
def check(m)
  rm = request.metadata(m)
  def f(x) =
    (fst(x),string.recode(out_enc="UTF-8",snd(x)))
  end
  metau = list.map(f, rm)
  if p.last(metau) < 3600. then
    log("was played, letting through")
  end
  p.add(metau)
  true
end

```

**Version details**

- OS: Debian 11
- Version: 2.1.2

**Install method**
Github actions packages

[1] For my example, it yields atrocity (iso=>utf8 ugliness): [
  "files/L'Ã¢ge idiot.mp3",
  1671149343.04
],

**Comment by vitoyucepi** [comment]
Hi @gilou,
I'm unable to reproduce this issue.
Could you please upload a minimal persistence log file that causes an error?

I don't know the exact name of the music file in your fs.
Here are the commands, that will give some clues.
1. `find . -name '*idiot.mp3' -type f`.
   This command will print the names of files that end with `idiot.mp3`.
2. `find . -name '*idiot.mp3' -type f | xxd`.
   The same, but the output will print in hex.

As far as I can reproduce this issue, the output of the commands looks like
```
./files/L'Ã¢ge idiot.mp3
00000000: 2e2f 6669 6c65 732f 4c27 c383 c2a2 6765  ./files/L'....ge
00000010: 2069 6469 6f74 2e6d 7033 0a               idiot.mp3.
```

**Comment by gilou** [comment]
Here's an [archive](https://github.com/savonet/liquidsoap/files/10245350/test.tar.gz) with the borked file and log, find is unhappy with the filename as well:

```
streams@liqv2:/media/liquidsoap/test/files$ find . -type f | xxd
00000000: 2e2f 3130 2d30 3820 4c27 e267 6520 6964  ./10-08 L'.ge id
00000010: 696f 742e 6d70 330a                      iot.mp3.

```

FS is ext4, file was uploaded from a system that does nasty things, but cleaning that up is not the issue at hand ;))

**Comment by vitoyucepi** [comment]
Got it!

```
4c  27  e2  67  65
 L   ' 342   g   e
```

https://github.com/savonet/liquidsoap/blob/b6d8db9b568306f104ede6176559e936c3a81389/libs/utils.liq#L81-L86

If you try printing `f`, then it will print `files/L'�ge.mp3`.
Ideally, after `let json.stringify t = f`, `t` should become `"files/L'\u00e2ge.mp3"`.
Instead, it preserves the invalid(in UTF-8) sequence.

https://github.com/savonet/liquidsoap/blob/b6d8db9b568306f104ede6176559e936c3a81389/libs/utils.liq#L58-L65

At load time `let json.parse ( parsed : [(string * float)]? ) = file.contents(null.get(persistency))` expects right utf-8 sequence, but instead it receives `\xe2`.
If you try to load this JSON into `jq`, then it will replace the invalid sequence with `\uFFFD` or `\xEF\xBF\xBD` instead of throwing an exception.

Here's a full reproduction

1. `mkdir files`
2. `ffmpeg -f lavfi -i "sine=frequency=1000:duration=1" "files/$(printf "\xe2").wav"`
3. *main.liq*
   ```ruby
   p = playlog(persistency="test.log")

   def check(m)
     rm = request.metadata(m)
     p.add(rm)
     true
   end

   pl = playlist("files", check_next=check)
   output.dummy(pl, fallible=true)
   ```
4. `liquidsoap main.liq`
5. wait some time
6. ctrl-c
7. `liquidsoap main.liq`

**Comment by vitoyucepi** [comment]
Another solution is to
```ruby
def hash(m)
  string.base64.encode(m["filename"])
end

p = playlog(persistency="test.log", hash=hash)
```

I wonder if there any other built-in hash functions.

@gilou also check [page 76](http://www.liquidsoap.info/book/book.pdf).

**Comment by gilou** [comment]
Yup, I do also have a workaround using the MD5 checksum of the file with files.digest, mentioned by David when he added the playlog(), and that's what  I need in some cases, but the default implementation seems a bit fragile (and my "solution" to just encode to UTF-8 was just because I had it laying around, it's probably not very clever as you noticed (but it works in that case :P)

**Comment by toots** [comment]
Thanks for the careful reporting and analysis y'all! `camomile` is supposed to guarantee that all strings we receive are in fact `utf8`, even with invalid chars replaced by the `replacement` character.

Your solution, thus, is doing what the liquidsoap internals are failing to do. Let me review the code to find out how we can make fix it on our end.
